### PR TITLE
Resolved PWA Inconsistencies #693

### PR DESCRIPTION
**Pull request by johnmgrant**
Closes https://github.com/RaenonX-PokemonSleep/pokemon-sleep-ui/issues/693

Improved VSCode development utils
* Updated ESLint to support formatting JSON files
* Added prettier format rules through VSCode prior to ESLint for code wrapping.
* Configured typescript override
* Added https experimental support

Configured dynamic manifest
* Modified Manifest to support localization
* Updated PWA to support offline experience
* Update device icon support

Fixed theme color inconsistency between system, site and PWA

### CODE REVIEW

File: src/utils/meta.ts [hunk 1]
```diff
@@ -1,25 +1,43 @@
...
   return {
-    metadataBase: process.env.NEXTAUTH_URL ? new URL(process.env.NEXTAUTH_URL) : null,
-    applicationName: siteName,
-    title: siteName,
-    description: siteDescription,
+    metadataBase,
+    alternates: {
+      canonical: '/',
+      languages: {
+        'en': '/en',
```
Comment (RaenonX) [inline src/utils/meta.ts:31]: Make string keys simply literal key (same coloring stuff mentioned above).

Also use `satisifes` to type this object `{[locale in Locale]: string}` so when new locale is added, then this object pops error.

File: src/utils/meta.ts [hunk 2]
```diff
@@ -81,11 +125,14 @@ export const generatePageMeta = ({key, values}: GeneratePageMetaOpts): GenerateM
 };
 
 type generatePageMetaFromStringOpts = {
-  t: (key: I18nMessageKeysOfNamespace<'UI.Metadata'>) => string,
-  title: string,
+  t: (key: I18nMessageKeysOfNamespace<'UI.Metadata'>) => string;
```
Comment (RaenonX) [inline src/utils/meta.ts:128]: `,` comma instead of `;` for consistency

File: src/utils/metadata.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import {SiteMetadata} from '@/types/next/manifest';
+import {getI18nTranslator} from '@/utils/i18n';
+import {GeneratePageMetaOpts} from '@/utils/meta';
+
+
+export const generateSiteMetadata = ({
+  key = 'Home.Title',
+  values = undefined,
+  locale = 'en',
+}: {
```
Comment (RaenonX) [inline src/utils/metadata.ts:11]: Type the parameter as `GenerateSiteMetadataOpts` instead.

File: src/utils/metadata.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
+import {Locale} from '@/types/next/locale';
+import {SiteMetadata} from '@/types/next/manifest';
+import {getI18nTranslator} from '@/utils/i18n';
+import {GeneratePageMetaOpts} from '@/utils/meta';
+
+
+export const generateSiteMetadata = ({
+  key = 'Home.Title',
+  values = undefined,
```
Comment (RaenonX) [inline src/utils/metadata.ts:9]: If `values` is `undefined` then it's gonna be `undefined`, so I don't think a default value assignment is needed here?

File: src/utils/metadata.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  locale = 'en',
+}: {
+  key?: GeneratePageMetaOpts['key'],
+  values?: GeneratePageMetaOpts['values'],
+  locale?: Locale,
+}): Promise<SiteMetadata> => {
+  return getI18nTranslator({locale, namespace: 'UI.Metadata'})
+    .then((t) => {
+      return {
```
Comment (RaenonX) [inline src/utils/metadata.ts:18]: If it's single line return, make it like this instead:
```typescript
.map(() => ({
  a: 'b',
})
```

File: src/utils/metadata.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        description: t('Site.Description'),
+        keywords: [
+          t('Site.Keywords.Pokemon'),
+          t('Site.Keywords.Pokémon'),
+          t('Site.Keywords.PokemonSleep'),
+          t('Site.Keywords.PokemonSleepGuide'),
+        ],
+        shortcuts: [
+          {
+            'name': t('Home.Title'),
```
Comment (RaenonX) [inline src/utils/metadata.ts:34]: Make this literal key instead of string key.

**Review by RaenonX — CHANGES_REQUESTED** [review]

File: src/utils/manifest.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    'short_name': shortName,
+    description,
+    'display': 'standalone',
+    'orientation': 'portrait',
+    'background_color': '#2a97d7',
+    'scope': metadataBase ? metadataBase.host : undefined,
+    'categories': ['productivity', 'utilities'],
+    shortcuts,
+    'display_override': ['window-controls-overlay', 'standalone', 'browser'],
+    // @todo: make this generative of images instead.
```
[the remark this answers is not in the document]
Comment (johnmgrant) [reply src/utils/manifest.ts:29]: Created #723

File: src/utils/manifest.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+}: GenerateManifestParams): Promise<Manifest> => {
+  const {
+    name,
+    shortName,
+    metadataBase,
+    description,
+    shortcuts,
+  } = await generateSiteMetadata({locale});
+
+  return {
```
[the remark this answers is not in the document]
Comment (johnmgrant) [reply src/utils/manifest.ts:16]: Updating ESLint to contain the following rule then.
```
  quote-props:
    - error
    - "as-needed"
    - keywords: true
      unnecessary: true
```

File: src/ui/base/navbar/darkMode/main.tsx [hunk 9]
```diff
@@ -10,7 +10,7 @@ import {useMounted} from '@/hooks/mounted';
 
 export const ThemeSwitcher = () => {
   const {mounted} = useMounted();
-  const {theme, setTheme} = useTheme();
+  const {resolvedTheme: theme, setTheme} = useTheme();
```
[the remark this answers is not in the document]
Comment (johnmgrant) [reply src/ui/base/navbar/darkMode/main.tsx:13]: `theme` initially defaults to 'system' on first user visit where as `resolveTheme` utilizes session object defined [here](https://github.com/RaenonX-PokemonSleep/pokemon-sleep-ui/blob/fa54381040dcf3e7dfb4668006fade9febf86bc5/src/ui/base/providers.tsx#L21-L22) to always return either `light` or `dark`. The result is that the site defaults to the users device settings and only if the theme icon is selected will that be overridden and persist to the PWA.

File: messages/ui-kr.json [hunk 10]
```diff
@@ -68,7 +68,13 @@
   "Metadata": {
     "Site": {
       "Name": "포슬립 위키",
-      "Description": "포켓몬 슬립 위키는 포켓몬의 수면 정보, 레시피, 식재료, 나무열매 뿐 아니라 팀 분석, 포켓몬 EXP 계산과 같은 툴을 포함한 다양한 정보와 기능을 제공합니다."
+      "Description": "포켓몬 슬립 위키는 포켓몬의 수면 정보, 레시피, 식재료, 나무열매 뿐 아니라 팀 분석, 포켓몬 EXP 계산과 같은 툴을 포함한 다양한 정보와 기능을 제공합니다.",
```
[the remark this answers is not in the document]
Comment (johnmgrant) [reply messages/ui-kr.json:71]: I think it just added a trailing comma here, but I reached out about the "Guide" translation.

**Comment by johnmgrant** [comment]
Closing this as it no longer would address #693. 

Breaking out the changes to individual issues.

**Pull request closed without merging**